### ISSUE TITLE
[Question] Making the Learning Rate Learnable for the Implicit Gradient Component

### ISSUE THREAD

**Issue by XindiWu** [metaopt/torchopt#163]
### Required prerequisites

- [X] I have read the documentation <https://torchopt.readthedocs.io>.
- [X] I have searched the [Issue Tracker](https://github.com/metaopt/torchopt/issues) and [Discussions](https://github.com/metaopt/torchopt/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [X] Consider asking first in a [Discussion](https://github.com/metaopt/torchopt/discussions/new).

### Questions

Hi @XuehaiPan @Benjamin-eecs @JieRen98,

I have been recently working with your codebase, and I've been trying to modify the [implicit gradient component ](https://github.com/metaopt/torchopt#implicit-gradient-ig) such that the learning rate becomes learnable as well.

As it stands now, the learning rate is a fixed parameter that we need to set before training. I have been trying to implement this change, but I have not been successful so far. Any guidance would be really helpful! Thank you very much for maintaining this project!

**Comment by XuehaiPan** [comment]
Hi @XindiWu, thanks for your attention and the question.

> As it stands now, the learning rate is a fixed parameter that we need to set before training.

I'm afraid that I do not quite understand your question. Do you mean the inner loop learning rate or the outer loop learning rate?

If you mean the inner loop learning rate, you can try:

```python
def stationary(params, meta_params, data):
    lr, *other_meta_params = meta_params
    # Stationary condition construction
    return stationary condition  # should be in the same structure as `params`, e.g., tuple of tensors

@torchopt.diff.implicit.custom_root(stationary)
def solve(params, meta_params, data):
    lr, *other_meta_params = meta_params
    # Solve the inner loop optimization
    # find a solution for `stationary(params, meta_params, data) == zeros`
    return solution  # should be in the same structure as `params`, e.g., tuple of tensors


lr_tensor = torch.tensor(learning_rate, requires_grad=True)
meta_params = (lr_tensor, *other_meta_params)
solution = solve(init_params, meta_params, data)
outer_loss = loss_fn(solution, meta_params, data)
outer_loss.backward()
lr_tensor.grad
```

You need to construct a `stationary` function that

$$
\frac{\partial \text{stationary}}{\partial \text{lr}} \ne 0
$$

------

If you mean the outer loop learning rate, that will not affect the inner loop's optimization process.

**Comment by XindiWu** [comment]
Got it, thank you! I wonder if there should be any changes in the @torchopt.diff.implicit.custom_root?

**Comment by XindiWu** [comment]
Thank you!

I also just checked the [doc](https://torchopt.readthedocs.io/en/latest/api/api.html#implicit-meta-gradient-module) and see the similar example as you gave before:

`def solve(self, batch, labels):
    parameters = tuple(self.parameters())
    optimizer = torch.optim.Adam(parameters, lr=1e-3)
    with torch.enable_grad():
        for _ in range(100):
            loss = self.objective(batch, labels)
            optimizer.zero_grad()
            # Only update the `.grad` attribute for parameters
            # and leave the meta-parameters unchanged
            loss.backward(inputs=parameters)
            optimizer.step()
    return self`

However, in the [colab notebook](https://colab.research.google.com/github/metaopt/torchopt/blob/main/tutorials/5_Implicit_Differentiation.ipynb#scrollTo=8d623b2f-48ee-4df6-a2ce-cf306b4c9067), the example is:

`def inner_solver(params, meta_params, data):
    # Initial functional optimizer based on TorchOpt
    x, y, fmodel = data
    optimizer = torchopt.sgd(lr=2e-2)
    opt_state = optimizer.init(params)
    with torch.enable_grad():
        # Temporarily enable gradient computation for conducting the optimization
        for i in range(100):
            pred = fmodel(params, x)
            loss = F.mse_loss(pred, y)  # compute loss

            # Compute regularization loss
            regularization_loss = 0.0
            for p1, p2 in zip(params, meta_params):
                regularization_loss += 0.5 * torch.sum(torch.square(p1.view(-1) - p2.view(-1)))
            final_loss = loss + regularization_loss

            grads = torch.autograd.grad(final_loss, params)  # compute gradients
            updates, opt_state = optimizer.update(grads, opt_state, inplace=True)  # get updates
            params = torchopt.apply_updates(params, updates, inplace=True)

    optimal_params = params
    return optimal_params`

I wonder what are the differences between these two?

**Comment by XuehaiPan** [comment]
> I wonder what are the differences between these two?

@XindiWu The first example comes from the docstring for the `torchopt.nn.ImplicitMetaGradientModule.solve` method. It's for the OOP API of the Implicit Gradient. The second example is using the functional API of the Implicit Gradient.

- OOP API: `torchopt.nn.ImplicitMetaGradientModule`
- functional API: `torchopt.diff.implicit.custom_root`

**Comment by XindiWu** [comment]
Got it, thank you!!

**Comment by XindiWu** [comment]
Awesome thank you again for the reply!!

**Comment by XuehaiPan** [comment]
Closing now. Feel free to ask more questions or ask for a reopening.
